llvm/lib/Target/RISCV/MCTargetDesc/RISCVAsmBackend.cpp

```diff
@@ -660,7 +698,16 @@ bool RISCVAsmBackend::addReloc(const MCFragment &F, const MCFixup &Fixup,
   if (IsResolved &&
       (getFixupKindInfo(Fixup.getKind()).Flags & MCFixupKindInfo::FKF_IsPCRel))
     IsResolved = isPCRelFixupResolved(Target.getAddSym(), F);
-  IsResolved = MCAsmBackend::addReloc(F, Fixup, Target, FixedValue, IsResolved);
+
+  if (!IsResolved) {
+    // Some Fixups require a vendor relocation, record it (directly) before we
+    // add the relocation.
+    if (std::optional<StringRef> VendorIdentifier =
```

Comment:
Combine the two functions and just call sth like `maybeAddVendorReloc(F, Fixup)`